core/src/main/java/kafka/server/share/DelayedShareFetch.java

```diff
@@ -149,15 +149,27 @@ public void onComplete() {
      */
     @Override
     public boolean tryComplete() {
-        log.trace("Try to complete the delayed share fetch request for group {}, member {}, topic partitions {}",
-            shareFetchData.groupId(), shareFetchData.memberId(),
-            shareFetchData.partitionMaxBytes().keySet());
+        // There can be multiple threads which might invoke tryComplete for same share fetch request
+        // hence check if delay share fetch request is already completed. If yes, return true.
+        // However, this check alone cannot guarantee that request is really completed. It is possible that
+        // tryComplete is invoked by multiple threads and state has yet not updated. Hence, we need to check
+        // the forceComplete response as well.
+        if (isCompleted()) {
```

Comment:
Can we also add unit tests for the conditions to verify this line and line 164

Reply:
Yeah, let me do that tomorrow. I have tested the runs manually and verified with 25 parallel share consumers and 25 million messages.

Reply:
I have added tests.